### PR TITLE
fix(broker): datasets get cached as not found permanently

### LINKED ISSUE #87: comet caches that it doesn't know datasets
The lru cache will also cache that a dataset is unknown and therefor also return an error after a previously unknown dataset is registered:
```
$ curl -d '{"ds_id": "foo"}' -X POST -H "Content-Type: application/json" http://localhost:12050/update-datasets
{"datasets":{},"result":"update-datasets: Dataset ID foo unknown to broker."}
$ curl -X POST -H 'Content-type: application/json' --data '{"hash":"foo","ds": {"dset": "b2da1fde9f6de323caacc6caf8eef780", "state": "foo", "is_root": true}}' http://localhost:12050/register-dataset
{"result":"success"} 
$ curl -d '{"ds_id": "foo"}' -X POST -H "Content-Type: application/json" http://localhost:12050/update-datasets
{"datasets":{},"result":"update-datasets: Dataset ID foo unknown to broker."}
```

### PR DESCRIPTION
**Pull request by nritsche**
Solved by making the lru_cached functions raise Exceptions instead of
returning None in case they don't find what was requested and telling
the lru_cache to not cache exceptions.

Closes #87

### CODE REVIEW

**Comment by nritsche** [comment]
It looks like the travis tests were successful, but it didn't report back: https://travis-ci.org/github/chime-experiment/comet/builds/732358709

And codecov doesn't seem to work properly yet.

File: comet/broker.py [hunk 1]
```diff
@@ -409,8 +411,19 @@ async def register_dataset(request):
...
+            except DatasetNotFoundError:
+                if dataset_valid and root is not None:
+                    # save the dataset
+                    await redis.execute("hset", "datasets", hash, json.dumps(ds))
+
+                    reply["result"] = "success"
+                    archive_ds = True
+
+                    # Notify anything waiting for this dataset to arrive
+                    signal_created(hash, "dataset", lock_datasets, waiting_datasets)
```
Comment (jrs65) [inline comet/broker.py:425]: What should happen if you get `DatasetNotFoundError` but you don't satisfy the `if` clause above? I appreciate this isn't covered in the original code either.

File: comet/broker.py [hunk 2]
```diff
@@ -644,25 +649,28 @@ async def get_dataset(ds_id, wait=True):
...
+    DatasetNotFoundError
+        If the dataset doesn't exist or waiting for the dataset timed out.
+
     """
     if wait:
         # Check if existing and wait if not
         found = await wait_for_dset(ds_id)
         if not found:
-            raise TimeoutError("Dataset {} not found: Timeout.".format(ds_id))
+            raise DatasetNotFoundError("Dataset {} not found: Timeout.".format(ds_id))
```
Comment (jrs65) [inline comet/broker.py:664]: Is there any reason you would still need to differentiate between a time out and the dataset actually not being found?

File: tests/test_datasets.py [hunk 3]
```diff
@@ -357,3 +359,34 @@ def test_register_start_dataset_automated(manager_and_dataset, broker):
...
+    ds_id = "doesntexist"
+
+    # Request an unknown dataset
+    with pytest.raises(BrokerError):
+        manager_low_timeout.get_dataset(ds_id)
+    state = manager_low_timeout.register_state(data={"foo": "bar"}, state_type="test")
+    assert state is not None
+    assert isinstance(state, State)
+
+    # Now register it (manurally, to set the ID)
```
Comment (jrs65) [inline tests/test_datasets.py:376]:
```suggestion
    # Now register it (manually, to set the ID)
```

**Review by jrs65 — APPROVED** [review]
Looks good.

File: comet/broker.py (hunk 1 above)
Comment (nritsche) [reply comet/broker.py:425]: result gets set to "dataset invalid", this branch gets skipped and the function returns. I agree it's not too readable.

**Review by jrs65 — APPROVED** [review]

**Pull request merged**